### PR TITLE
feat: adds Product selection when using ticket

### PR DESCRIPTION
**Pull request by marfavi**
Add support for using concrete products when claiming a ticket:
- Upgraded ticket use flow to accommodate the selection of menu items explicitly.
- The last selected menu item is cached locally for each ticket (cache cleared on logout).

other:
- Simplified the Product feature file structure.
- Add improvements to network request handling by adopting a TaskEither type for better composability.
- Added package Hive for local storage
- Rename some files and refactor several imports across the codebase
- Fixed various minor bugs, tightened type constraints, and made use of pattern matching & fpdart types.

### CODE REVIEW

**Review by TTA777 — COMMENTED** [review]
There are frankly so many changes, I don't follow all of it.

With that said, only some smaller comments, where I would like some thoughts from you.

Also, are you sure that caching the last used menu item on swiping will have the intended outcome? I'm thinking that people will tend to be lazy, and just use whatever it was already on when swiping. This would be a shame, since it would basically make the data we get from it worthless. Is this something you have considered?

File: lib/features/product/product_repository.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import 'package:coffeecard/core/network/network_request_executor.dart';
+import 'package:coffeecard/features/product/product_model.dart';
+import 'package:coffeecard/generated/api/coffeecard_api_v2.swagger.dart';
+import 'package:collection/collection.dart';
+import 'package:fpdart/fpdart.dart';
+
+/// Repository for fetching products.
+///
+/// The repository caches all products fetched from the API.
+class ProductRepository {
```
[the remark this answers is not in the document]
Comment (marfavi) [reply lib/features/product/product_repository.dart:11]: Simplification. "Flutter clean architecture" code just _feels bad_ to use in general, be it adding new functionality to a feature or understanding already-written code. One example of use cases feeling bad is that they are a class with no inheritance and a implements a single .call method (with no contract enforcing this), which means that in your code, you do this:

```dart
class FetchAllProducts {
    ...
    SomeReturnType call() { ... } // will be called if you call the class
}
```

In another file:

```dart
FetchAllProducts(); // <- what return type does this have?
```

- Why are we calling a class?
- What is the return type of `FetchAllProducts()`? Hovering over the call doesn't work.
- If a use case simply implements .call, why not use a global function instead?

Our NetworkRequestExecutor used the same trick back in the day; now we explicitly call `NetworkRequestExecutor.execute(...)` or `NetworkRequestExecutor.executeAsTask(...)` instead.

Using a repository instead makes me understand the code much better.

File: lib/features/ticket/presentation/widgets/tickets_section.dart [hunk 2]
```diff
@@ -31,128 +31,108 @@ class TicketSection extends StatelessWidget {
...
-                // Remove the swipe overlay
-                Navigator.of(context, rootNavigator: true).pop();
-              }
-
-              LoadingOverlay.show(context).ignore();
+            // Implicitly cast to TicketsLoaded
+            // (the superclass of the Using, Used and UseError states)
+            //
+            // This is already checked in the listenWhen function, so this
+            // is purely for type safety
```
[the remark this answers is not in the document]
Comment (marfavi) [reply lib/features/ticket/presentation/widgets/tickets_section.dart:40]: Yes specifically for type promotion. The compiler should be angry with the switch case if you remove the check

**Comment by marfavi** [comment]
> There are frankly so many changes, I don't follow all of it.
> 
> With that said, only some smaller comments, where I would like some thoughts from you.
> 
> Also, are you sure that caching the last used menu item on swiping will have the intended outcome? I'm thinking that people will tend to be lazy, and just use whatever it was already on when swiping. This would be a shame, since it would basically make the data we get from it worthless. Is this something you have considered?

It was implemented to help people feel like not much extra work is being done. We can ask baristas about how precise customer swipes 1-2 weeks in with this feature enabled.

**Review by TTA777 — APPROVED** [review]

**Pull request merged**